core/include/gnuradio-4.0/Port.hpp

```diff
@@ -845,76 +960,48 @@ class DynamicPort {
+inline constexpr TagPredicate auto defaultEOSTagMatcher = [](const Tag &tag, Tag::signed_index_type readPosition) noexcept {
+    auto eosTagIter = tag.map.find(gr::tag::END_OF_STREAM);
+    if (eosTagIter != tag.map.end() && eosTagIter->second == true) {
+        if (tag.index >= readPosition || tag.index < 0) {
+            return true;
+        }
     }
+    return false;
+};
+} // namespace detail
 
-    // at least one tag is present -> if tag is not on the first tag position read up to the tag position
-    const auto tagData           = port.tagReader().get();
-    const auto readPosition      = port.streamReader().position();
-    const auto future_tags_begin = std::ranges::find_if(tagData, [&readPosition](const auto &tag) noexcept { return tag.index > readPosition + 1; });
+inline constexpr std::optional<std::size_t>
+nSamplesToNextTagConditional(const PortLike auto &port, detail::TagPredicate auto &predicate, Tag::signed_index_type readOffset) {
+    const std::span<const Tag> tagData = port.tagReader().template get<false>();
+    if (!port.isConnected() || tagData.empty()) [[likely]] {
+        return std::nullopt; // default: no tags in sight
+    }
+    const Tag::signed_index_type readPosition = port.streamReader().position();
 
-    if (future_tags_begin == tagData.begin()) {
-        const auto        first_future_tag_index   = static_cast<std::size_t>(future_tags_begin->index);
-        const std::size_t n_samples_until_next_tag = readPosition == -1 ? first_future_tag_index : (first_future_tag_index - static_cast<std::size_t>(readPosition) - 1UZ);
-        return n_samples_until_next_tag;
+    // at least one tag is present -> if tag is not on the first tag position read up to the tag position, or if the tag has a special 'index = -1'
+    const auto firstMatchingTag = std::ranges::find_if(tagData, [&](const auto &tag) { return predicate(tag, readPosition + readOffset); });
+    if (firstMatchingTag != tagData.end()) {
+        return static_cast<std::size_t>(std::max(firstMatchingTag->index - readPosition, Tag::signed_index_type(0))); // Tags in the past will have a negative distance -> deliberately map them to '0'
     } else {
-        return 0;
+        return std::nullopt;
     }
 }
 
-constexpr std::size_t
-samples_to_eos_tag(const PortLike auto &port) {
-    if (!port.isConnected() || port.tagReader().available() == 0) [[likely]] {
-        return std::numeric_limits<std::size_t>::max(); // default: no tags in sight
-    }
-    const auto tags    = port.tagReader().get();
-    const auto readPos = port.streamReader().position();
-
-    // find the smallest index of EOS tag, if no EOS tag is present then index = std::numeric_limits<std::size_t>::max()
-    std::size_t result = std::numeric_limits<std::size_t>::max();
-    // TODO: 'const auto tag' without reference is a workaround for gcc compiler
-    //  by unknown reasons only copy works, reference leads to unvalidated tag.map.end() iterator
```

Comment:
Is this problem no longer valid because the logic was changed and only the first Tag is searched for?

Reply:
Yes, this is fixed now. The primary issue seemed to be UB in the circular/mmap-ed buffer that does not work (by C++ design/standard) for non-trivially copyable data types. I will also inform the GR 3.X team (ping to @mormj) that this is also likely a UB bug in the existing GR 3.X and probably may reemerge also for fundamental types as the compiler will perform better/more aggressive optimisations (the double-mmap-ed version's end is nigh, and the <5% throughput benefit compared to the clean std-conforming implementation doesn't warrant the hassle much longer).